[event-epoll.c] Check for slots in next table if all slots are used

black-dragon74 opened the pull request:
Simplified the while loop to remove logically dead code.

Updates: #1060
CID: 1469575

Signed-off-by: black.dragon74 <nickk.2974@gmail.com>

mykaul commented on libglusterfs/src/event-epoll.c:
```diff
@@ -78,22 +78,16 @@ __event_slot_alloc(struct event_pool *event_pool, int fd,
 retry:
 
-    while (i < EVENT_EPOLL_TABLES) {
+    if (i < EVENT_EPOLL_TABLES) {
```
nit (and perhaps not related to this patch) - can we just use table_idx and get rid of i? Why do we need both?

xhernandez replied:
This condition is always true and should be removed (`i` will always be `< EVENT_EPOLL_TABLES` unless it's defined as 0, but it shouldn't).

The other question is why we have `EVENT_EPOLL_TABLES` defined as 1024 if we only use the first one ?

Maybe the problem is not the dead code but that in some case we should keep iterating the remaining tables ?

With `EVENT_EPOLL_SLOTS` also defined as 1024, it means that we may only have 1024 active epoll entries, which translates into a limit of ~1000 connections (1 epoll slot per connection). For big volumes, or for servers with many bricks using brick-mux, that could be a problem.

mykaul replied:
> This condition is always true and should be removed (`i` will always be `< EVENT_EPOLL_TABLES` unless it's defined as 0, but it shouldn't).
> 
> The other question is why we have `EVENT_EPOLL_TABLES` defined as 1024 if we only use the first one ?

I've asked this @ https://github.com/gluster/glusterfs/issues/3228

> 
> Maybe the problem is not the dead code but that in some case we should keep iterating the remaining tables ?
> 
> With `EVENT_EPOLL_SLOTS` also defined as 1024, it means that we may only have 1024 active epoll entries, which translates into a limit of ~1000 connections (1 epoll slot per connection). For big volumes, or for servers with many bricks using brick-mux, that could be a problem.

mykaul commented:
What's the next step here?

black-dragon74 commented:
> What's the next step here?

Waiting on @xhernandez 

Should I open the PR mentioned above (related to counter variable removal)?

Regards

mykaul commented:
> > What's the next step here?
> 
> Waiting on @xhernandez
> 
> Should I open the PR mentioned above (related to counter variable removal)?

Just push https://github.com/black-dragon74/glusterfs/commit/fc413d2037c4445dec4828d37f57177c509d3da4 into this one, IMHO.

- or we could do it with the next round of improvements to this code. 
> 
> Regards

xhernandez commented:
> > What's the next step here?
> 
> Waiting on @xhernandez

I'm not very familiar with this part of the code, so I don't know if we need to modify the exit condition from the loop of simply remove the loop. It seems to me that we should keep the loop, but I can't tell for sure. @mohit84 probably you have more knowledge about what's best here.

mohit84 commented:
> > > What's the next step here?
> > 
> > 
> > Waiting on @xhernandez
> 
> I'm not very familiar with this part of the code, so I don't know if we need to modify the exit condition from the loop of simply remove the loop. It seems to me that we should keep the loop, but I can't tell for sure. @mohit84 probably you have more knowledge about what's best here.

Here we have two loops, the job of first loop is to identify a table that has space to save the slots, a table can have at max
 1024 slots. The next loop is to identify free slots. The current code of the function is not correct, we changed the code from the
 patch c2af6523c58771ad4956322f525367e216918c53. We need to change the condition to work it correctly. In case if all slots are used we need to check next free table, here it is returning -1 so in case if any gluster process need to consume more than 1024 slots it will not work fine.

```diff
diff --git a/libglusterfs/src/event-epoll.c b/libglusterfs/src/event-epoll.c
index 9e61bc7a61..3ef378de1d 100644
--- a/libglusterfs/src/event-epoll.c
+++ b/libglusterfs/src/event-epoll.c
@@ -81,10 +81,12 @@ retry:
     while (i < EVENT_EPOLL_TABLES) {
         table = event_pool->ereg[i];
         if (table) {
-            if (table->slots_used == EVENT_EPOLL_SLOTS)
-                return -1;
-            else
+            if (table->slots_used == EVENT_EPOLL_SLOTS) {
+                i++;
+                continue;
+            } else {
                 break; /* break out of the loop */
+            }
         } else {
             table = __event_newtable(event_pool, i);
             if (!table)
@@ -96,6 +98,9 @@ retry:
         i++;
     }
 
+    if (!table)
+        return -1;
+
     table_idx = i;
```

black-dragon74 committed:
[event-epoll.c] Check for next free table if all slots are used

If all epoll slots are used in the current table look for the free slots
in the next table. In rare cases, if a gluster process required more than
1024 slots the previous code returned -1. This patch addresses the same.

Signed-off-by: black-dragon74 <niryadav@redhat.com>

black-dragon74 commented:
The extra check for table being NULL, suggested at https://github.com/gluster/glusterfs/pull/3268#issuecomment-1062690647,  is not needed as it is already handled by the conditional blocks above itself?

Additionally, removed the extraneous counter variable `i` as it was merely a mirror of `table_idx`.

Please correct me if wrong.

CC: @mykaul @xhernandez @mohit84 

Regards

black-dragon74 commented:
/run regression

mohit84 commented on libglusterfs/src/event-epoll.c:
```diff
@@ -70,34 +70,32 @@ static int
 
-    table_idx = i;
-
     for (j = 0; j < EVENT_EPOLL_SLOTS; j++) {
```
I think before checking free slot we need to check table_idx should be less than EVENT_EPOLL_TABLES.

black-dragon74 replied:
```c
if (table_idx < EVENT_EPOLL_TABLES) { 
    for (j = 0 ....)
    /* Rest of the code */
} else {
    // Shouldn't we print something to the logs here?
    // It is highly unlikely that we will not find free slots but for edge cases?
    // I remember that 1024 tables is way too much and we will likely find free slots in one of them
    // Should I just add the if block and skip else part then?
}
```
Thoughts?

mohit84 replied:
Yes we can but it would not be beneficial because parent function is already log a message.

black-dragon74 replied:
Okay. Got it.

xhernandez commented on libglusterfs/src/event-epoll.c:
```diff
@@ -70,60 +70,60 @@ static int
-            memset(&table[j], 0, sizeof(table[j]));
-            table[j].fd = fd;
-            table[j].gen = gen + 1;
+    if (table_idx < EVENT_EPOLL_TABLES) {
```
This is not correct. If `table_idx` is equal to `EVENT_EPOLL_TABLES`, which means that there are no tables available, `j` will be 0 and the check in line 124 will incorrectly assume that a slot has been found.

I think it would be better to just check the error condition (i.e. `table_idx >= EVENT_EPOLL_TABLES`) and return -1 in that case. Otherwise continue normally.

black-dragon74 replied:
Done.

black-dragon74 committed:
[event-epoll.c] Ensure that table_idx < EVENT_EPOLL_TABLES

before looking for free slots.

Signed-off-by: black-dragon74 <niryadav@redhat.com>

mohit84 commented:
/run regression
